<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Error Handling</title>
</head>
<body>
    
    <script> 

        class MathUtil {
            static sum(...numbers) {

                if (numbers.length === 0) {
                    throw new Error(`Total Parameter harus lebih dari 0`)
                }

                let total = 0
                for (const number of numbers) {
                    total += number
                }

                return total
            }
        }

        try { // tempat meletakkan kode yang mungkin salah
            console.log(MathUtil.sum())
            console.log("Huda")
        } catch(error) { // tempat menangkap / tereksekusinya kode jika terjadi error 
            console.log(`Terjadi error ${error.message}`) // properti bawaan dari Error untuk mengambil value
        } finally {
            console.log(`Program selesai`) // finally, akan di eksekusi baik program terjadi error atau tidak
        }

        console.log(MathUtil.sum(1, 1, 1))

        // try catch ; menangani agar program tidak langsung berhenti saat terjadi error

        class Counter {
            #counter = 0

            next() {
                try {
                    return this.#counter
                } finally { // setelah try, tidak wajib untuk diikuti oleh catch. ini dilakukan jika yakin di try kode sudah benar dan tidak ingin di tangani disini jika terjadi error 
                    this.#counter++
                }
            }
        }

        const counter = new Counter()
        console.log(counter.next())
        console.log(counter.next())
        console.log(counter.next())
        console.log(counter.next())
        console.log(counter.next())
        console.log(counter)

    </script>

</body>
</html>